src/app/Shared/Services/Login.service.tsx

```diff
@@ -0,0 +1,174 @@
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+import { Observable, ObservableInput, of, ReplaySubject } from 'rxjs';
+import { fromFetch } from 'rxjs/fetch';
+import { catchError, first, map, tap } from 'rxjs/operators';
+
+export class LoginService {
+
+  private readonly TOKEN_KEY: string = 'token';
+  private readonly METHOD_KEY: string = 'method';
+  private readonly USER_KEY: string = 'user';
+  private readonly token = new ReplaySubject<string>(1);
+  private readonly authMethod = new ReplaySubject<string>(1);
+  private readonly login = new ReplaySubject<void>(1);
+  private readonly logout = new ReplaySubject<void>(1);
+  private readonly username = new ReplaySubject<string>(1);
+  readonly authority: string;
+
+  constructor() {
+    let apiAuthority = process.env.CRYOSTAT_AUTHORITY;
+    if (!apiAuthority) {
+      apiAuthority = '';
+    }
+    this.authority = apiAuthority;
+    this.token.next(this.getCacheItem(this.TOKEN_KEY));
+    this.authMethod.next(this.getCacheItem(this.METHOD_KEY));
+    this.username.next(this.getCacheItem(this.USER_KEY));
+  }
+
+  checkAuth(token: string, method: string): Observable<boolean> {
+
+    token = this.useCacheItemIfAvailable(this.TOKEN_KEY, token);
+
+    return fromFetch(`${this.authority}/api/v1/auth`, {
+      credentials: 'include',
+      mode: 'cors',
+      method: 'POST',
+      body: null,
+      headers: this.getAuthHeaders(token, method),
+    })
+    .pipe(
+      map(response => {
+        if (!this.authMethod.isStopped) {
+          this.authMethod.next(response.ok ? method : (response.headers.get('X-WWW-Authenticate') || ''));
+        }
+        return response.ok;
+      }),
+      catchError((e: Error): ObservableInput<boolean> => {
+        window.console.error(JSON.stringify(e));
+        this.authMethod.complete();
+        return of(false);
+      }),
+      first(),
+      tap(v => {
+        if (v) {
+          this.authMethod.next(method);
+          this.authMethod.complete();
+          this.token.next(token);
+          this.setCacheItem(this.TOKEN_KEY, token);
+          this.setCacheItem(this.METHOD_KEY, method);
+        }
+      })
+    );
+  }
+
+  getToken(): Observable<string> {
+    return this.token.asObservable();
+  }
+
+  getAuthHeaders(token: string, method: string): Headers {
+    const headers = new window.Headers();
+    if (!!token && !!method) {
+      headers.set('Authorization', `${method} ${token}`)
+    }
+
+    return headers;
+  }
+
+  getAuthMethod(): Observable<string> {
+    return this.authMethod.asObservable();
+  }
+
+  getUsername(): Observable<string> {
+    return this.username.asObservable();
+  }
+
+  isAuthenticated(): boolean {
```

Comment:
It might be better here to return an `Observable<boolean>`. You can just use `of` for now.

Reply:
If I use `of()` to convert `isAuthenticated` to an Observable, this useState hook would require an initial boolean value. 
AppLayout.tsx, line 76
```
const [showUserIcon, setShowUserIcon] = React.useState(serviceContext.login.isAuthenticated());
```
I ended up adding a custom hook to update the useState hook whenever the value of `isAuthenticated` is updated. Please let me know what you think.
```
const showUserIcon = useObservable(serviceContext.login.isAuthenticated());
```

Reply:
I think defaulting it to `false` at first is fine, and then updating it as needed once the actual value is available. Since that's currently just coming from sessionStorage it's local and synchronous and so the update can be done very quickly, but in the future when we have server-side sessions, it will have to be asynchronous and defaulting to `false` and waiting for a network request to complete would be required anyway.